improv_ble_client/improv_ble_client.py

```diff
@@ -0,0 +1,469 @@
+                    raise UnexpectedDisconnect from err
+                raise Disconnected(self._disconnect_reason) from err
+            except ImprovError:
+                self._disconnect(DisconnectReason.ERROR)
+                raise
+
+    async def _ensure_connected(self) -> None:
+        """Ensure connection to device is established."""
+        if self._connect_lock.locked():
+            _LOGGER.debug(
+                "%s: Connection already in progress, waiting for it to complete; "
+                "RSSI: %s",
+                self.name,
+                self.rssi,
+            )
+        if self._client and self._client.is_connected:
+            self._reset_disconnect_timer()
+            return
+        async with self._connect_lock:
+            # Check again while holding the lock
+            if self._client and self._client.is_connected:
+                self._reset_disconnect_timer()
+                return
+            _LOGGER.debug("%s: Connecting; RSSI: %s", self.name, self.rssi)
+            client = await establish_connection(
+                BleakClientWithServiceCache,
+                self._ble_device,
+                self.name,
+                self._disconnected,
+                use_services_cache=False,  # True
+                ble_device_callback=lambda: self._ble_device,
+            )
+            _LOGGER.debug("%s: Connected; RSSI: %s", self.name, self.rssi)
+
+            self._client = client
+            self._disconnect_reason = None
+            self._reset_disconnect_timer()
+
+            _LOGGER.debug(
+                "%s: Subscribe to notifications; RSSI: %s", self.name, self.rssi
+            )
+            await client.start_notify(
+                CHARACTERISTIC_UUID_ERROR, self._notification_handler
+            )
+            await client.start_notify(
+                CHARACTERISTIC_UUID_RPC_RESULT, self._rpc_result_handler
+            )
+            await client.start_notify(
+                CHARACTERISTIC_UUID_STATE, self._notification_handler
+            )
+
+    def _raise_if_not_connected(self) -> None:
+        """Raise if the connection to device is lost."""
+        if self._client and self._client.is_connected:
+            self._reset_disconnect_timer()
+            return
+        raise NotConnected
+
+    def _reset_disconnect_timer(self) -> None:
+        """Reset disconnect timer.
+
+        If the disconnect timer expires, disconnect from the device.
+        """
+
+        async def _disconnect() -> None:
+            """Execute disconnect request."""
+            _LOGGER.debug(
+                "%s: Disconnecting after timeout of %s",
+                self.name,
+                DISCONNECT_DELAY,
+            )
+            await self._execute_disconnect(DisconnectReason.TIMEOUT)
+
+        def _schedule_disconnect() -> None:
+            self._disconnect_timer = None
+            asyncio.create_task(_disconnect())
```

Comment:
Hold a strong reference to the background task to ensure it doesn't get garbage collected

Reply:
Added. Maybe the tasks should be awaited in the reset function?

Reply:
This is how we fixed pyhap 
https://github.com/ikalchev/HAP-python/blob/f95cd15a427f05bb51240a7b2e1df8aab0bfb20e/pyhap/util.py#L183

Note that we have to use ensure_future there because we support older python

Reply:
OK, that's exactly the implementation which is added by https://github.com/home-assistant-libs/py-improv-ble-client/pull/2/commits/3fa1d3c6597806f4cf4edd723f224a5729b9fb84, so I think it's fine now 👍